services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py

```diff
@@ -204,80 +241,94 @@ async def _scale_up_cluster(app: FastAPI, pending_tasks: list[Task]) -> None:
     list_of_ec2_instances = await ec2_client.get_ec2_instance_capabilities(
         app_settings.AUTOSCALING_EC2_INSTANCES
     )
+    # get the task in larger cpu resources to smaller
+    pending_tasks.sort(
+        key=lambda t: utils_docker.get_max_resources_from_docker_task(t).cpus
+    )
+
+    # some instances might be able to run several tasks
+    list_of_instance_to_tasks: list[tuple[EC2Instance, list[Task]]] = []
     for task in pending_tasks:
+        if _try_assigning_task_to_instances(task, list_of_instance_to_tasks):
```

Comment:
So you dont scale if by your calculations the cluster looks like it has enough resources to handle the pending tasks?

Usually, if all is well, tasks are pending for, my gutfeeling says, <5 seconds. If a task is pending for more than >10 seconds, what we have seen in the past is one of:
- Missing resources ( --> invoke autoscaling)
- Cant schedule due to constraints ( --> most likely a dev error)
- Issues with dask-schedular having 3 replica but 1 requested, some were pending (let's call it "unstable swarm")
- Further "unstable swarm" situations


While I think this code behaviour here is fine in principle, I invite you (or other reviewers) to think about what autoscaling should do if it detects pending tasks that can theoretically be scheduled, but de-facto are not even after many seconds (since we have encountered this already a few times). I guess at the very least we would want an email alert, since most likely we are in "unstable swarm" territory then. But also, I am not sure if the autoscaler should (or can) explicitly check how long a task has been pending and act differently depending on it.

Just a comment for now, came to mind.

Reply:
so the autoscaler only checks for the tasks with missing resources, which are specifically labelled to be auto-scaled, none other.